File: core/src/main/scala/kafka/api/ControlledShutdownRequest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.api

import java.nio.ByteBuffer

import kafka.common.TopicAndPartition
import kafka.api.ApiUtils._
import kafka.network.{RequestOrResponseSend, RequestChannel}
import kafka.network.RequestChannel.Response
import kafka.utils.Logging
import org.apache.kafka.common.protocol.{ApiKeys, Errors}

object ControlledShutdownRequest extends Logging {
  val CurrentVersion = 1.shortValue
  val DefaultClientId = ""

  def readFrom(buffer: ByteBuffer): ControlledShutdownRequest = {
    val versionId = buffer.getShort
    val correlationId = buffer.getInt
    val clientId = if (versionId > 0) Some(readShortString(buffer)) else None
    val brokerId = buffer.getInt
    new ControlledShutdownRequest(versionId, correlationId, clientId, brokerId)
  }

}

case class ControlledShutdownRequest(versionId: Short,
                                     correlationId: Int,
                                     clientId: Option[String],
                                     brokerId: Int)
  extends RequestOrResponse(Some(ApiKeys.CONTROLLED_SHUTDOWN_KEY.id)){

  if (versionId > 0 && clientId.isEmpty)
    throw new IllegalArgumentException("`clientId` must be defined if `versionId` > 0")

  def writeTo(buffer: ByteBuffer) {
    buffer.putShort(versionId)
    buffer.putInt(correlationId)
    clientId.foreach(writeShortString(buffer, _))
    buffer.putInt(brokerId)
  }

  def sizeInBytes: Int = {
    2 + /* version id */
      4 + /* correlation id */
      clientId.fold(0)(shortStringLength) +
      4 /* broker id */
  }

  override def toString: String = {
    describe(true)
  }

  override  def handleError(e: Throwable, requestChannel: RequestChannel, request: RequestChannel.Request): Unit = {
    val errorResponse = ControlledShutdownResponse(correlationId, Errors.forException(e).code, Set.empty[TopicAndPartition])
    requestChannel.sendResponse(new Response(request, new RequestOrResponseSend(request.connectionId, errorResponse)))
  }

  override def describe(details: Boolean = false): String = {
    val controlledShutdownRequest = new StringBuilder
    controlledShutdownRequest.append("Name: " + this.getClass.getSimpleName)
    controlledShutdownRequest.append("; Version: " + versionId)
    controlledShutdownRequest.append("; CorrelationId: " + correlationId)
    controlledShutdownRequest.append(";ClientId:" + clientId)
    controlledShutdownRequest.append("; BrokerId: " + brokerId)
    controlledShutdownRequest.toString()
  }
}

File: clients/src/main/java/org/apache/kafka/common/requests/ControlledShutdownRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.common.requests;

import org.apache.kafka.common.message.ControlledShutdownRequestData;
import org.apache.kafka.common.message.ControlledShutdownResponseData;
import org.apache.kafka.common.protocol.ApiKeys;
import org.apache.kafka.common.protocol.ByteBufferAccessor;
import org.apache.kafka.common.protocol.Errors;

import java.nio.ByteBuffer;

public class ControlledShutdownRequest extends AbstractRequest {

    public static class Builder extends AbstractRequest.Builder<ControlledShutdownRequest> {

        private final ControlledShutdownRequestData data;

        public Builder(ControlledShutdownRequestData data, short desiredVersion) {
            super(ApiKeys.CONTROLLED_SHUTDOWN, desiredVersion);
            this.data = data;
        }

        @Override
        public ControlledShutdownRequest build(short version) {
            return new ControlledShutdownRequest(data, version);
        }

        @Override
        public String toString() {
            return data.toString();
        }
    }

    private final ControlledShutdownRequestData data;

    private ControlledShutdownRequest(ControlledShutdownRequestData data, short version) {
        super(ApiKeys.CONTROLLED_SHUTDOWN, version);
        this.data = data;
    }

    @Override
    public ControlledShutdownResponse getErrorResponse(int throttleTimeMs, Throwable e) {
        ControlledShutdownResponseData data = new ControlledShutdownResponseData()
                .setErrorCode(Errors.forException(e).code());
        return new ControlledShutdownResponse(data);
    }

    public static ControlledShutdownRequest parse(ByteBuffer buffer, short version) {
        return new ControlledShutdownRequest(new ControlledShutdownRequestData(new ByteBufferAccessor(buffer), version),
            version);
    }

    @Override
    public ControlledShutdownRequestData data() {
        return data;
    }
}
